torch/_refs/__init__.py

```diff
@@ -4647,6 +4648,48 @@ def triu_indices(
     )
 
 
+@register_decomposition(torch.ops.aten.bucketize)
+@out_wrapper()
+def bucketize(
+    a: TensorLikeType,
+    boundaries: TensorLikeType,
+    *,
+    out_int32: bool = False,
+    right: bool = False,
+):
+
+    out_dtype = torch.int32 if out_int32 else torch.int64
+    n_boundaries = boundaries.shape[-1]
+
+    # can't use int32 as indexes, so we have to do all computations with int64 and convert at the end
+    start = torch.zeros(a.shape, device=a.device, dtype=torch.int64)
+    end = start + n_boundaries
+    # Max depth of the binary search
+    niters = int(math.log2(n_boundaries)) + 1
```

Comment:
Let's add a comment describing the algorithm

Reply:
Sure, done.

Reply:
Is the number of iterations correct given that this now populates the original tensors with the first iteration?

Reply:
Put another way, shouldn't it be possible to never go through the loop with the current algorithm?

Reply:
I see what you mean, I rewrote it slightly to work like this.
Note that it will only skip the loop if boundaries has a single entry, so not sure how much it will matter in practice, but it's there.